File: bindings/android/scion-http3-android/src/main/kotlin/com/anapaya/scion/http3/internal/UniffiHttp3Backend.kt
// Copyright 2026 Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.anapaya.scion.http3.internal

import com.anapaya.scion.http3.uniffi.HttpRequest
import com.anapaya.scion.http3.uniffi.HttpResponse
import com.anapaya.scion.http3.uniffi.defaultClientConfig
import com.anapaya.scion.http3.uniffi.ScionHttp3Client as FfiClient

/**
 * [Http3Backend] over the generated bindings.
 *
 * Nothing but delegation and construction lives here. Everything that decides anything is above this
 * class, which is what lets the tests above it run without the native library this one loads.
 */
internal class UniffiHttp3Backend(
    private val client: FfiClient,
) : Http3Backend {
    override suspend fun execute(request: HttpRequest): HttpResponse = client.execute(request)

    override suspend fun warmUp(url: String) {
        client.warmUp(url)
    }

    override fun reset() {
        client.reset()
    }

    override fun setAuthToken(token: String) {
        client.setAuthToken(token)
    }

    override suspend fun shutdown() {
        client.shutdown()
    }

    override fun close() {
        client.close()
    }
}

/**
 * Builds a real backend.
 *
 * Blocking, and the caller is expected to be on a dispatcher that tolerates that: the constructor
 * installs a cryptography provider and starts the runtime that carries every request, and reading
 * the platform's trust anchors touches the disk.
 */
internal class UniffiHttp3BackendFactory(
    private val trustStore: SystemTrustStore,
) : Http3BackendFactory {
    override fun create(settings: ClientSettings): Http3Backend {
        val config =
            settings.applyTo(
                // The stack's own defaults, asked for by URL because that is the one setting it has
                // no default for.
                base = defaultClientConfig(settings.endhostApiUrl),
                trust = settings.trust.toFfi(trustStore),
            )
        return UniffiHttp3Backend(FfiClient(config))
    }
}

File: bindings/apple/scion-http3-swift/Sources/ScionHTTP3/Internal/UniffiHttp3Backend.swift
// Copyright 2026 Anapaya Systems
internal import ScionHTTP3Uniffi

/// `Http3Backend` over the generated bindings.
final class UniffiHttp3Backend: Http3Backend {
    private let client: FfiClient

    init(client: FfiClient) {
        self.client = client
    }

    func newCancelHandle() -> any Http3CancelHandle {
        CancelHandle()
    }

    func execute(
        _ request: HttpRequest, cancel: any Http3CancelHandle
    ) async throws -> HttpResponse {
        guard let handle = cancel as? CancelHandle else {
            preconditionFailure("a cancel handle from another backend was passed to this one")
        }
        return try await client.executeCancellable(request: request, cancel: handle)
    }

    func warmUp(_ url: String) async throws {
        try await client.warmUp(url: url)
    }

    func reset() {
        client.reset()
    }

    func setAuthToken(_ token: String) throws {
        try client.setAuthToken(token: token)
    }

    func shutdown() async {
        await client.shutdown()
    }
}

extension CancelHandle: Http3CancelHandle {}

/// Builds a real backend. This is the first call into the native library a client makes: it starts
/// the runtime that carries every request.
let uniffiBackendFactory: Http3BackendFactory = { settings in
    let config = settings.applyTo(defaultClientConfig(endhostApiUrl: settings.endhostApiUrl))
    return UniffiHttp3Backend(client: try FfiClient(config: config))
}
